### ISSUE TITLE
Additional metrics for block delays

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#2528]
[attestation production routine]: https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/beacon_chain/src/beacon_chain.rs#L1254-L1428
[`BEACON_BLOCK_IMPORTED_SLOT_START_DELAY_TIME`]: https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/beacon_chain/src/metrics.rs#L629-L632
[`BEACON_BLOCK_HEAD_SLOT_START_DELAY_TIME`]: https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/beacon_chain/src/metrics.rs#L633-L636
[`BEACON_BLOCK_GOSSIP_SLOT_START_DELAY_TIME`]: https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/network/src/metrics.rs#L464-L467
[in the `BeaconChain::fork_choice` function]: https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/beacon_chain/src/beacon_chain.rs#L2947-L2951

## Le Problème

There is a well-known issue where validators frequently "miss" the head/target votes (i.e., vote for a block that is not canonical). Assuming that our [attestation production routine] is solid (I believe it is), these missed votes occur because we haven't managed to import and set the correct block as our head before the attestation is created (4s after the slot starts). There's two reasons why this might happen:

- Our peers sent us the block too late (i.e., later than 4s from slot start).
- We got the block in time, but it took us too long to import it and set it as the head.

This information is not new, I've been chasing this problem for quite a while. In my previous efforts, I added some metrics:

1. [`BEACON_BLOCK_GOSSIP_SLOT_START_DELAY_TIME`]: The delay between the start of the slot and when we saw it on the gossip network.
1. [`BEACON_BLOCK_IMPORTED_SLOT_START_DELAY_TIME`]: The delay between the start of the slot and when we verified the block and imported it to the DB.
1. [`BEACON_BLOCK_HEAD_SLOT_START_DELAY_TIME`]: The delay between the start of the slot and when we set it as the canonical head.

These metrics work well on their own, however I'm unable to link them together to answer this important question:

*"When a block is set as the canonical head too late, is it because we saw the block too late? Or, is it because it took us too long to import it? Or, did we manage to import it in time but were too slow to set it as the head?"*

The answer to this question might help us understand if we need to focus on spreading blocks around the network faster (i.e., gossip propagation times), *or*, if we need to focus on speeding up our block import or fork choice times.

So, what's stopping us from linking those metrics together?

When we set metric (3) [in the `BeaconChain::fork_choice` function], we've completely lost the time when we saw the block on the network and when we finished verifying it. Importing the block to the DB is a *completely separate function* to running the fork choice algorithm to find and set the canonical head.

We need to be able to make the network/import delay information available to the `BeaconChain::fork_choice` function.

## La Solution Potentielle

Create a new cache and stick it to the `BeaconChain`. It would likely look something like this:

```rust
struct Timestamps {
  observed: Option<Duration>,
  imported: Option<Duration>,
  set_as_head: Option<Duration>
}

type BlockRoot = Hash256;

struct BlockDelays(HashMap<BlockRoot, Timestamps>);
```
We set `observed` when we see the block via:

- Network (RPC)
- Network (Gossip)
- HTTP API

We set `imported` somewhere around [here](https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/beacon_chain/src/beacon_chain.rs#L2461-L2468).

We set `set_as_head` (and report some new metrics/logs) somewhere around [here](https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/beacon_chain/src/beacon_chain.rs#L2947-L2951).

### Considérations Supplémentaires

Whenever we add a cache we need to be sure that we bound the size of it and prune it. For this cache, I think we only need to track blocks from the last couple of epochs. That's probably only 64ish blocks and relatively minuscule compared to everything else. We could trigger a prune on [`BeaconChain::per_slot_task`](https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/beacon_chain/src/beacon_chain.rs#L3122-L3130) and just prune blocks that are `n` slots older than the current slot.

Due to some valid edge cases, we might have the case where set the head and we don't have timing info. That's totally fine, perhaps we just log a `debug!` log.

Presently, metric (1) only tracks gossip blocks. We should probably track blocks from the RPC too. Any block from the network goes via the [`BeaconProcessor`](https://github.com/sigp/lighthouse/blob/unstable/beacon_node/network/src/beacon_processor/mod.rs). You can [see](https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/network/src/beacon_processor/mod.rs#L328) that we track the `seen_timestamp` for gossip blocks but [not for rpc blocks](https://github.com/sigp/lighthouse/blob/12fe72bd372b6ffc788572aa0b3c559961a97a55/beacon_node/network/src/beacon_processor/mod.rs#L429-L431). I'm not sure how hard this is, but we can *probably* follow that function trail upstream and take a timestamp at the earliest possible time and thread it through the `BeaconProcessor`. If this gets hard, feel free to reach out to me :)

**Comment by paulhauner** [comment]
#2533 is somewhat related to this.

**Comment by macladson** [comment]
Resolved in #2546
